dynamic/internal/shim/grpcutil/translate.go

```diff
@@ -94,7 +105,207 @@ func Translate[
 		return tmp, err
 	}
 	if pulumiTFDebugGRPCLogs != "" {
-		log.Response = v.String()
+		jsonOut, jsonErr := json.Marshal(v)
+		if jsonErr != nil {
+			err = jsonErr
+		}
+		log.Response = string(jsonOut)
 	}
 	return m(v), nil
 }
+
+type LogReplayProvider struct {
+	methodLogs map[string][]grpcLog
+}
+
+var _ = tfplugin6.ProviderClient(LogReplayProvider{})
+
+func getMethodFromFullName(fullName string) string {
+	splitName := strings.Split(fullName, ".")
+	return splitName[len(splitName)-1]
+}
+
+func NewLogReplayProvider(name, version, url, logs string) LogReplayProvider {
+	grpcLogs := make([]grpcLog, 0)
```

Comment:
It's generally not a great idea to distinguish between `[]T(nil)` and `[]T{}`. Since Go forces us to deal with `[]T(nil)` as the zero value of `[]T`, we try to stick to that for simplicity.

Another example is here:

https://github.com/pulumi/pulumi/blob/bd63383a89b6a6761d757af494b96ae12ee76e55/sdk/go/common/slice/slice.go#L17-L25